pwd_guess.py

```diff
@@ -965,31 +990,62 @@ def training_step(self, x_all, y_all, w_all):
             x_val, y_val, sample_weight=w_val)
         return (train_loss, train_accuracy, test_loss, test_accuracy)
 
-    def train_model_generation(self):
+    def early_stopping(self,smooth_loss, serializer):
+        stop=False
+        if self.min_loss_early_stopping > smooth_loss:
+            self.min_loss_early_stopping = smooth_loss
+            serializer.save_model(self.model_to_save)
+            self.poor_batches_early_stopping = 0
+        elif self.poor_batches_early_stopping < self.config.early_stopping_patience:
+            self.poor_batches_early_stopping +=1
+        else:
+            stop = True
+        return stop
+
+    def train_model_generation(self, serializer=None):
+        assert(not(self.config.early_stopping and serializer is not None)), "Need to specify serializer with early_stopping"
         self.chunk = 0
         self.pwd_list.reset()
         accuracy_accum = []
+        smoothened_loss = collections.deque(maxlen= self.config.chunk_print_interval)
         x_all, y_all, w_all = self.next_train_set_as_np()
         chunk = 0
+        early_stop = False
         while len(x_all) != 0:
             assert len(x_all) == len(y_all)
-            tr_loss, _, te_loss, te_acc = self.training_step(
+            tr_loss, tr_acc, te_loss, te_acc = self.training_step(
                 x_all, y_all, w_all)
             accuracy_accum += [(len(x_all), te_acc)]
+            smoothened_loss.append((len(x_all),te_loss))
+            if self.config.tensorboard:
+                self.write_log(self.train_log_names,[tr_loss, tr_acc], self.cumulative_chunks)
+                self.write_log(self.test_log_names,[te_loss,te_acc], self.cumulative_chunks)
             if chunk % self.config.chunk_print_interval == 0:
+                instances_smoothened = map(lambda x: x[0], smoothened_loss)
+                loss_smoothened = sum(map(lambda x: x[0] * x[1], smoothened_loss)) / sum(instances_smoothened)
```

Comment:
Why are you multiplying the loss times the number of training instances here. I would have thought that the loss was already aggregated across all training instances. Is that wrong? If you think this is correct, maybe add a comment explaining why multiplying here is best. Also, using a named tuple here would be more readable instead of having to figure out what the 0 and 1 indexes are.